pkg/controller/history/controller_history_test.go

```diff
@@ -77,30 +77,24 @@ func TestRealHistory_ListControllerRevisions(t *testing.T) {
 	}
 	ss1 := newStatefulSet(3, "ss1", types.UID("ss1"), map[string]string{"foo": "bar"})
 	ss2 := newStatefulSet(3, "ss2", types.UID("ss2"), map[string]string{"goo": "car"})
-	sel1, err := metav1.LabelSelectorAsSelector(ss1.Spec.Selector)
-	if err != nil {
-		t.Fatal(err)
-	}
-	sel2, err := metav1.LabelSelectorAsSelector(ss2.Spec.Selector)
-	if err != nil {
-		t.Fatal(err)
-	}
-	ss1Rev1, err := NewControllerRevision(ss1, parentKind, sel1, rawTemplate(&ss1.Spec.Template), 1, nil)
+	sel1 := labels.Set(ss1.Spec.Template.Labels).AsSelector()
```

Comment:
It's always safe to use the StatefulSet's selector, because we ensure through validation that it selects the pod template labels. Since we now apply the pod template labels to revisions too, we can also expect that the StatefulSet's selector will match those revisions.

On the other hand, generating a selector purely from the template labels will not always give the right result. Here is a contrived example of that:

```yaml
...
metadata:
  name: ss1
spec:
  selector:
    matchLabels:
      component: mything
    matchExpressions:
    - {key: role, operator: DoesNotExist}
  template:
    metadata:
      labels:
        component: mything
...
---
...
metadata:
  name: ss2
spec:
  selector:
    matchLabels:
      component: mything
      role: test
  template:
    metadata:
      labels:
        component: mything
        role: test
```

If you generate a selector directly from the pod template labels, ss1 would end up also selecting ss2's revisions, whereas ss1's actual selector means that it will avoid selecting ss2's pods. We want the behavior for selecting pods and revisions to be consistent with each other, so we respect the user's intention.

I called this example contrived because it's probably a bad idea for the user to do this. However, the API allows it, so we need to make sure we do the right thing.